### PR TITLE
LV-Generate-Data Finished

### LINKED ISSUE #4: [Core, Data] Add Product
А Product should have the following properties:  
Id, Title, Description, Brand, Price, Quantity, and Product category

add the db configuration as well

### LINKED ISSUE #5: [Data] Generate test data for products
Use data generation tool like https://www.mockaroo.com/ to build xml with as much as possible products 

Add an XML reader 
Seed the data to the db on app start

### PR DESCRIPTION
**Pull request by LVarsaminov**

### CODE REVIEW

File: SportsGoods.Api/SportsGoods.Api.csproj [hunk 1]
```diff
@@ -9,6 +9,11 @@
   <ItemGroup>
     <PackageReference Include="MediatR" Version="12.2.0" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
```
Comment (clangelov) [inline SportsGoods.Api/SportsGoods.Api.csproj]: No unit testing stuff in this project @LVarsaminov

File: SportsGoods.App/Services/ProductService.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                var brandElement = element.Element("Brand");
+                var priceElement = element.Element("Price");
+                var quantityElement = element.Element("Quantity");
+                var productCategoryElement = element.Element("ProductCategory");
+
+                Guid productId;
+
+                if (idElement != null && Guid.TryParse(idElement.Value, out productId))
+                {
+                    if (!string.IsNullOrEmpty(titleElement?.Value) &&
```
Comment (clangelov) [inline SportsGoods.App/Services/ProductService.cs]: Extract validation in a separate method to improve readability @LVarsaminov

File: SportsGoods.App/SportsGoods.App.csproj [hunk 3]
```diff
@@ -9,6 +9,7 @@
   <ItemGroup>
     <PackageReference Include="MediatR" Version="12.2.0" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
```
Comment (clangelov) [inline SportsGoods.App/SportsGoods.App.csproj]: No testing nuget packages in this project

File: SportsGoods.Core/Interfaces/IApplicationDbContext.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.EntityFrameworkCore;
+using SportsGoods.Core.Models;
+
+namespace SportsGoods.Core.Interfaces
+{
+    public interface IApplicationDbContext
+    {
+        DbSet<Product> Products { get; }
+        //Task<int> SaveChangesAsync();
```
Comment (clangelov) [inline SportsGoods.Core/Interfaces/IApplicationDbContext.cs]: if something is commented, then you don't need it, therefore no need to be in the PR

File: SportsGoods.Core/Interfaces/IProductRepository.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SportsGoods.Core.Interfaces
+{
+    public interface IProductRepository
+    {
+        IQueryable<Product> Products { get; }
```
Comment (clangelov) [inline SportsGoods.Core/Interfaces/IProductRepository.cs]: No IQueryable in Repositories 

In this post I am going to make the case why returning IQueryable from a Repository Pattern is an anti-pattern and why you should not do it.

https://weblogs.asp.net/dotnetstories/repository-iqueryable

File: SportsGoods.Core/SportsGoods.Core.csproj [hunk 6]
```diff
@@ -10,6 +10,7 @@
     <PackageReference Include="MediatR" Version="12.2.0" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
     <PackageReference Include="Microsoft.Extensions.Identity.Stores" Version="8.0.3" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
```
Comment (clangelov) [inline SportsGoods.Core/SportsGoods.Core.csproj]: no testing nugets over here as well

File: SportsGoods.Data/DAL/ApplicationDbContext.cs [hunk 7]
```diff
@@ -2,45 +2,32 @@
...
-                     Password = "password2"
-                 },
-                  new Administrator
-                  {
-                      Id = new Guid(),
-                      Email = "ThirdAdmin@abv.bg",
-                      Password = "password3"
-                  }
-            );
         }
```
Comment (clangelov) [inline SportsGoods.Data/DAL/ApplicationDbContext.cs]: if it's commented out then it's not needed

File: SportsGoods.Data/SportsGoods.Data.csproj [hunk 8]
```diff
@@ -11,6 +11,7 @@
     <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="8.0.3" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="8.0.3" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
```
Comment (clangelov) [inline SportsGoods.Data/SportsGoods.Data.csproj]: no need for that package over here

File: UnitTests/Tests/ProductServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        [Test]
+        public void SeedProductsFromXml_NonExistentXmlFile_ThrowsFileNotFoundException()
+        {
+            var nonExistentXmlFilePath = "nonexistent.xml";
+
+            var productService = new ProductService(_context);
+
+            Assert.Throws<FileNotFoundException>(() =>
```
Comment (clangelov) [inline UnitTests/Tests/ProductServiceTest.cs]: I believe you can use https://docs.nunit.org/articles/nunit/writing-tests/assertions/classic-assertions/Assert.ThrowsAsync.html

File: SportsGoods.App/Services/ProductService.cs (hunk 2 above)
Comment (LVarsaminov) [reply SportsGoods.App/Services/ProductService.cs]: Done

File: SportsGoods.Core/Interfaces/IApplicationDbContext.cs (hunk 4 above)
Comment (LVarsaminov) [reply SportsGoods.Core/Interfaces/IApplicationDbContext.cs]: Done (removed)

File: SportsGoods.Data/DAL/ApplicationDbContext.cs (hunk 7 above)
Comment (LVarsaminov) [reply SportsGoods.Data/DAL/ApplicationDbContext.cs]: Done (removed).

File: SportsGoods.Data/SportsGoods.Data.csproj (hunk 8 above)
Comment (LVarsaminov) [reply SportsGoods.Data/SportsGoods.Data.csproj]: Done(removed).

File: UnitTests/Tests/ProductServiceTest.cs (hunk 9 above)
Comment (LVarsaminov) [reply UnitTests/Tests/ProductServiceTest.cs]: Done (refactored)

File: SportsGoods.Api/SportsGoods.Api.csproj (hunk 1 above)
Comment (LVarsaminov) [reply SportsGoods.Api/SportsGoods.Api.csproj]: Done (removed)

File: SportsGoods.App/SportsGoods.App.csproj (hunk 3 above)
Comment (LVarsaminov) [reply SportsGoods.App/SportsGoods.App.csproj]: Done (removed)

File: SportsGoods.Core/SportsGoods.Core.csproj (hunk 6 above)
Comment (LVarsaminov) [reply SportsGoods.Core/SportsGoods.Core.csproj]: Done (removed)

File: SportsGoods.Core/Interfaces/IProductRepository.cs (hunk 5 above)
Comment (LVarsaminov) [reply SportsGoods.Core/Interfaces/IProductRepository.cs]: Done. Refactored IProductRepository and ProductRepository. The interface no longer implements IQueryable.

File: SportsGoods.Core/Models/Product.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        public Guid Id { get; set; }
+        public string Title { get; set; } = string.Empty;
+        public string? Description { get; set; } = string.Empty;
+        public string Brand { get; set; } = string.Empty;
+        public double Price { get; set; }
+        public int Quantity { get; set; }
+        public string ProductCategory { get; set; } = string.Empty;
+
+        public static explicit operator Product(Task<Product?> v)
```
Comment (clangelov) [inline SportsGoods.Core/Models/Product.cs]: I think we don't need this one any longer @LVarsaminov

Comment (LVarsaminov) [reply SportsGoods.Core/Models/Product.cs]: Done (removed)

**Review by clangelov — APPROVED** [review]
looking good

**Pull request merged**